Using SQLAlchemy and SQL queries, I extract EU industry production data from the PostgreSQL database where I previously stored it.

In [1]:
import sqlalchemy  # Package for accessing SQL databases via Python

# Connect to database
engine = sqlalchemy.create_engine("postgresql://postgres:xfkLVeMj@localhost/production")
con = engine.connect()

There is just one table in the database:

In [2]:
print(engine.table_names())

['industry_production']


Let's see how the table is structured. This can be done using table reflections (see SQLAlchemy documentation: http://docs.sqlalchemy.org/en/latest/core/reflection.html). First, a MetaData object has to be created. The instance has then to be passed to the Table class, along with the table name and some autoload arguments to associate it with engine:

In [3]:
metadata = sqlalchemy.MetaData()
industry_production = sqlalchemy.Table('industry_production', metadata, autoload=True, autoload_with=engine)

The columns of the table can now be displayed with the columns attribute:

In [4]:
print(industry_production.columns)

['industry_production.indic_bt', 'industry_production.nace_r2', 'industry_production.s_adj', 'industry_production.unit', 'industry_production.geo\\time', 'industry_production.2017M08 ', 'industry_production.2017M07 ', 'industry_production.2017M06 ', 'industry_production.2017M05 ', 'industry_production.2017M04 ', 'industry_production.2017M03 ', 'industry_production.2017M02 ', 'industry_production.2017M01 ', 'industry_production.2016M12 ', 'industry_production.2016M11 ', 'industry_production.2016M10 ', 'industry_production.2016M09 ', 'industry_production.2016M08 ', 'industry_production.2016M07 ', 'industry_production.2016M06 ', 'industry_production.2016M05 ', 'industry_production.2016M04 ', 'industry_production.2016M03 ', 'industry_production.2016M02 ', 'industry_production.2016M01 ', 'industry_production.2015M12 ', 'industry_production.2015M11 ', 'industry_production.2015M10 ', 'industry_production.2015M09 ', 'industry_production.2015M08 ', 'industry_production.2015M07 ', 'industry_prod

The column names contain the table name and the actual column name, separated by a dot.
How many columns are there? Let's find out by getting the length of the column name list:

In [5]:
print(len(industry_production.columns))

781


Okay, how can I make sense of the column names now? Let's first look at columns 6 to 781: They contain the year (four digits) and the month (two digits), separated by an "M". So each month between August 2017 and January 1953 has its own column, and we go back in time from left to right. Annoyingly, there is a trailing space in the column names, which is an artefact from reading in the dataset.

The first five columns are more cryptic. Unfortunately, the EU Open Data Portal seems to hide information on the data format well. Before trying to find out more about the column names, let's see what kind of values are stored in them.

To access the data in the table, I can use queries in the SQL language. The basic syntax starts with the "SELECT" keyword, followed by the column names that are being requested, then the "FROM" keyword and the table name (because there are usually more than one table in the database). To select all columns, I use the star operator. Without specifying any filter, all the rows in the table are selected. This means that to retrieve the whole table, the query would be "SELECT * FROM industry_production". However, the table is rather clunky, so let me use a slightly modified query instead, for counting the number of rows in the table. This is done by applying the "count()" function, like this:

In [6]:
result = con.execute('SELECT count(*) FROM industry_production')
print(result.fetchall())

[(19197,)]


So there are 19197 rows in the table.

How do the values in the table look like? To get some idea, let me select the values for the first seven columns (i.e., the five "cryptic" columns, presumably some index) plus the first two data columns. I select the columns explicitly by name. Due to the use of the backslash character, "geo\\time" has to be enclosed in quotation marks. The same is the case for the data columns as they start with a numeric character (which actually is bad SQL practice) and end with a space, which has to be explicitly enclosed. I also limit the output to the first five rows, using the LIMIT keyword:

In [7]:
result = con.execute('SELECT indic_bt, nace_r2, s_adj, unit, "geo\\time", "2017M08 ", "2017M07 " FROM industry_production LIMIT 5')
print(result.fetchall())

[('PROD', 'B', 'CA', 'I10', 'AT', ': ', '108.8 p'), ('PROD', 'B', 'CA', 'I10', 'BA', '122.1 ', '115.3 '), ('PROD', 'B', 'CA', 'I10', 'BE', ': ', '102.0 p'), ('PROD', 'B', 'CA', 'I10', 'BG', ': ', '98.0 p'), ('PROD', 'B', 'CA', 'I10', 'CY', ': ', '60.8 p')]


The first five columns apparently have string values, where only the fifth column varies over the first five rows. This lets me suspect that the values are categorical, i.e. are chosen from a finite set of distinct elements. How many different elements are there for each column? Let's find out with queries using the DISTINCT keyword before the column name, and use one query per column:

In [8]:
print(con.execute('SELECT DISTINCT indic_bt FROM industry_production').fetchall())

[('PROD',)]


So the first column only has one value! Looking at the rather confusing metadata documentation on the EU Open Data Portal website (http://ec.europa.eu/eurostat/cache/metadata/en/sts_esms.htm), the "indic_bt" column can be related to Section 3.4 therein. "PROD" seems to stand for "production" and might be used to distinguish this dataset from other data products (turnover, volume of sales, etc.). But this info is already contained in the table name and thus this column is redundant.

What about the "nace_r2" column? Let me look at the alphabetically sorted list:

In [9]:
categories = con.execute('SELECT DISTINCT nace_r2 FROM industry_production').fetchall()
print(sorted(categories))

[('B',), ('B-D',), ('B-D_F',), ('B-D_X_FOOD',), ('B05',), ('B051',), ('B052',), ('B06',), ('B061',), ('B062',), ('B07',), ('B071',), ('B072',), ('B08',), ('B081',), ('B089',), ('B0891',), ('B0892',), ('B0893',), ('B0899',), ('B09',), ('B_C',), ('B_C_X_FD_MIG_NRG',), ('B_C_X_MIG_NRG',), ('C',), ('C10',), ('C10-C12',), ('C101',), ('C1011',), ('C1012',), ('C1013',), ('C102',), ('C103',), ('C1031',), ('C1032',), ('C1039',), ('C104',), ('C1041',), ('C1042',), ('C105',), ('C1051',), ('C1052',), ('C106',), ('C1061',), ('C1062',), ('C107',), ('C1071',), ('C1072',), ('C1073',), ('C108',), ('C1081',), ('C1082',), ('C1083',), ('C1084',), ('C1085',), ('C1086',), ('C1089',), ('C109',), ('C1091',), ('C1092',), ('C10_C11',), ('C11',), ('C1101',), ('C1102',), ('C1103',), ('C1104',), ('C1105',), ('C1106',), ('C1107',), ('C12',), ('C13',), ('C13-C15',), ('C131',), ('C132',), ('C133',), ('C139',), ('C1391',), ('C1392',), ('C1393',), ('C1394',), ('C1395',), ('C1396',), ('C1399',), ('C13_C14',), ('C14',), 

Here I get many cryptic categories that start with the letters "B", "C", "D" or "E", followed by other letters that are separated with a dash or underscore, or numerical characters. The metadata documentation can help here as well (again after some skimming). Section 3.2 and 3.3 say: "NACE Rev.2 classification (Statistical Classification of Economic Activities in the European Community) is used for all the STS indicators, except Industrial Import Prices (...)" and "INDUSTRY:
The indicators in this sector cover economic activities listed in sections B to E of NACE (B-Mining and quarrying, C-Manufacturing, D-Electricity, gas, steam and air conditioning supply, E-Water supply; sewerage, waste management and remediation activities". So the letters are a code for the industry sector.

Clicking on the "NACE Rev.2" link produces a list of these sectors, with further information about the numerical characters. For example, under letter "B" we find the following info: "Mining activities are classified into divisions, groups and classes on the basis of the principal mineral produced. Divisions 05, 06 are concerned with mining and quarrying of fossil fuels (coal, lignite, petroleum, gas); divisions 07, 08 concern metal ores, various minerals and quarry products."

Some categories start with "MIG". The metadata info gives some info on those ones in Section 18.6:
"industrial production: European aggregates and data for LU: total, MIGs (intermediate goods, energy, capital goods, durable consumer goods, non-durable consumer goods)".

To summarize, the "nace_r2" values are codes for industry (sub-)sectors.

Let's continue with the third column, "s_adj":

In [10]:
print(con.execute('SELECT DISTINCT s_adj FROM industry_production').fetchall())

[('NSA',), ('CA',), ('SCA',)]


Here, there are only three different categories. The column name seems to be an abbreviation for "seasonal adjustment" (again Section 18.6). 'NSA' apparently stands for "No seasonal adjustment", 'CA' for "Calendar adjustment", and 'SCA' for "Calendar and seasonal adjustment" (This can also be seen from the drop-down menus in the online interactive table version of the dataset, see http://appsso.eurostat.ec.europa.eu/nui/show.do?dataset=sts_inpr_m&lang=en).

The next column is "unit":

In [11]:
print(con.execute('SELECT DISTINCT unit FROM industry_production').fetchall())

[('I10',), ('PCH_SM',), ('PCH_PRE',)]


Again, there are three distinct values. The interactive table is helpful once more: 'I10' means that the value is an index that is normalized to 100 in the year 2010, 'PCH_SM' is the percentage change compared to the same month in the previous year, and 'PCH_PRE' is the percentage change compared to the previous month.

This leaves me with the fifth column, "geo\\time":

In [12]:
print(con.execute('SELECT DISTINCT "geo\\time" FROM industry_production').fetchall())

[('IE',), ('ES',), ('EA19',), ('MT',), ('BE',), ('AT',), ('BA',), ('EL',), ('NL',), ('CY',), ('DK',), ('SI',), ('BG',), ('CZ',), ('TR',), ('LT',), ('RS',), ('IT',), ('LU',), ('SK',), ('NO',), ('PT',), ('FI',), ('PL',), ('DE',), ('EE',), ('HU',), ('ME',), ('SE',), ('UK',), ('MK',), ('FR',), ('LV',), ('RO',), ('EU28',), ('HR',)]


These values can be recognized as the geo codes for the 28 EU member states, plus a few other countries (like 'TR' for Turkey) and aggregated values for the Euro currency area ('EA19') and the whole European Union ('EU28').

Hence the "geo" part of the "geo\\time" column name. The backslash separates the "geo" part (rows) from the "time" part (columns). Since the column name should just describe the column values, the "\\time" part should not be in it.

Now that I understand the meaning of the five index columns: How many unique combinations of these columns are there? This I can find out using the DISTINCT keyword once again, but applied to all five columns at once. Using the AS aliasing keyword, I give the name "unique_index_columns" to the result of this query. "unique_index_columns" is a table itself, and so I can apply the count syntax that I used earlier to the "unique_index_columns" table to obtain the number of unique combinations (this is an example for a nested query):

In [13]:
result = con.execute('SELECT count(*) FROM \
                     (SELECT DISTINCT indic_bt, nace_r2, s_adj, unit, "geo\\time" FROM industry_production) \
                     AS unique_index_columns')
print(result.fetchall())

[(19197,)]


So there are 19197 unique combinations of the values in the five columns, which matches the total number of rows in the table. This is consistent with the pandas dataframe from which the industry_production table was constructed (see previous project), where the five columns made up a multi-level index. Note that the "indic_bt" column is superfluous though, since it always has the same value.

At the end, let me now extract a part of this dataset from the database. I will select the manufacturing ("C") production data index for France, adjusted for calendar and season. These conditions can be specified in SQL with the WHERE keyword and combined with the AND keyword. Note that I have to escape the single quotes with a backslash to tell Python that they are part of the query string and not marking the beginning or end of the string:

In [21]:
query = 'SELECT * FROM industry_production WHERE nace_r2=\'C\' AND s_adj=\'SCA\' AND unit=\'I10\' AND "geo\\time"=\'FR\''
result = con.execute(query)
print(result.fetchall())

[('PROD', 'C', 'SCA', 'I10', 'FR', ': ', '104.4 ', '104.2 ', '105.1 ', '103.0 ', '104.5 ', '101.6 ', '102.3 ', '102.8 ', '104.1 ', '101.3 ', '101.6 ', '103.4 ', '100.5 ', '100.9 ', '101.8 ', '102.5 ', '100.5 ', '102.0 ', '103.2 ', '102.7 ', '102.3 ', '103.2 ', '102.7 ', '103.1 ', '100.2 ', '102.2 ', '101.2 ', '100.8 ', '101.3 ', '100.4 ', '100.3 ', '101.7 ', '98.5 ', '99.8 ', '100.8 ', '98.7 ', '100.9 ', '99.6 ', '97.9 ', '100.5 ', '100.0 ', '100.7 ', '99.7 ', '100.3 ', '100.7 ', '100.4 ', '100.0 ', '99.4 ', '100.7 ', '100.3 ', '101.3 ', '101.3 ', '99.1 ', '99.6 ', '99.4 ', '100.1 ', '99.1 ', '99.3 ', '100.7 ', '103.3 ', '101.7 ', '100.7 ', '100.8 ', '101.2 ', '102.7 ', '100.9 ', '102.3 ', '103.6 ', '104.8 ', '102.4 ', '102.3 ', '102.5 ', '103.8 ', '102.3 ', '106.5 ', '103.0 ', '104.3 ', '105.4 ', '104.8 ', '102.6 ', '101.5 ', '100.6 ', '101.0 ', '99.0 ', '101.0 ', '100.1 ', '100.2 ', '99.6 ', '99.6 ', '97.3 ', '97.7 ', '97.4 ', '98.8 ', '97.3 ', '98.3 ', '96.8 ', '96.0 ', '96.8 ', '95

The query result can be directly written into a pandas dataframe with the "read_sql_query()" method:

In [24]:
import pandas as pd
df = pd.read_sql_query(query, engine)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Columns: 781 entries, indic_bt to 1953M01
dtypes: object(781)
memory usage: 6.2+ KB


Pandas has taken care to also extract the column names.